### DISCUSSION TITLE
Can mongock do data migration to convert data types?

### DISCUSSION THREAD

**Discussion by tsanjoto** [flamingock/mongock#52] [Q&A] [answered]
## Description

Can mongock do data migration to convert data types? and is there an example of this? The example listed is just prepending data or multiply values, but nothing on changing data types.

My class Box to be migrated

```
Class Box{
    @Id
    @BsonProperty("id")
    private ObjectId id;

    String weight;
}
```

the current state of the data is populated with 
```
[  {weight: "100kg", id: ...},
   {weight: "80lbs", id: ...}  ]
```
and then the Box class is updated into

```
Class Box{
    @Id
    @BsonProperty("id")
    private ObjectId id;
    Double weight;
    String weightUnitName;
}
```

How do I achieve this in mongock? or is it better just to run a js script?

Currently I'm stuck on `mongoTemplate.find( new Query(), Box.class)` error out since the property `weight` stored is a string, and when I'm querying it with the new schema, java complains it can't convert to double properly due to '100kg'.

is there a way to cleanly do this or do I have to put it in a new property/variable?

## PRIORITY

NORMAL

### Mongock
- Mongock version: 5.0.29
- Modules involved: mongock-springboot, mongodb-springdata-v3-driver
- How Mongock is used: annotation

**Comment by dieppa** [comment, marked as answer]
Hello @tsanjoto ,

Of course Mongock can do it, it's actually one of the main advantages of it.

But before getting into it, you need to be aware that you are not only trying to change the data type. You are trying to change the actual value as well, which makes it a bit more complex(regardless the tool you use)

Also what you are trying to do is a it paradoxical 😅 . You are trying to change a class definition by using the class itself.

In this case what you should be doing it's to descend one level of abstraction and use MongoDatabase with document, instead of the `Box` class with MongoTemplate, as you mentioned it expects a Double, but you have a string stored in your database.

Something like this:
```java

    FindIterable<Document> boxes = database.getCollection("boxCollection").find(new Document());
    //parser your boxes iterable to a Box.class java collection(you obviously need to map values like String(100kg) to Double(100))
    // and then perform an update to override the items in your collection
```

I hope it helps
